### DISCUSSION TITLE
Tradeoffs of higher simulation BW?

### DISCUSSION THREAD

**Discussion by Sarajevo67** [thliebig/openEMS-Project#136] [Q&A Usage]
I noticed dramatic reduction of excitation signal length (in timesteps), inversely proportional to increase of excitation bandwidth. Despite it is initially counterintuitive I understand that it has solid foundation in physics. Also I understand that one of potential tradeoffs, because numerical integration mass is spreaded, might be a loss of accuracy. My questions:

1. How we can assess and mitigate that loss of accuracy?
2. What are other, higher excitation BW, tradeoffs?

Best Regards,

DD

**Comment by gadiLhv** [comment]
Well... The increase in **BW** for a shorter time signal is actually rather intuitive.

Consider this: A very short pulse (maybe even a Gaussian pulse?) will change rapidly in time. Rapid changes => higher frequencies. It's rather basic signal processing stuff (or harmonic analysis) and I am neglecting all the interesting material.

The signals you usually use for S-param analysis are modulated Gaussians. I can't plot their frequency response right here, but I will, in the next few days.

So here are the two limitations, off the top of my head.

1. **Lowest frequency**: If you fiddle with the modulated Gaussians a bit, you would see that you can't extend the frequency band indefinitely, as the pulse will hit 0 at some point. You would want it to decay properly (at least 90%) before hitting f=0.

2. **Highest frequency**: This is more intuition than fact checking. The highest frequency should require a short enough time step. This immediately creates a trade-off because of simulation time.

Cheers

**Comment by gadiLhv** [comment]
## **1. Additions the frequency domain answer**
Here are some necessary additions to my [previous comment](https://github.com/thliebig/openEMS-Project/discussions/136#discussioncomment-7527786).

The frequency response of a Gaussian pulse, 
$\large s(t) = e^{-\frac{1}{2}\({\frac{t - \Delta_t}{\sigma_t/\sqrt{8}}}\)^2}$
with $\sigma_t = 1/BW$, is as demonstrated here
![image](https://github.com/thliebig/openEMS-Project/assets/3908295/18e496d4-9cee-42b4-bd60-d29c87bae816)

Now, there is a whole thing here I recall. Gaussians, although they decay very close to zero, have infinite time response. In the classical sense, this function cannot be transformed using a Fourier Transform (**FT**). Something about Not withstanding the conditions of the L2 function space, hence cannot be considered as a distribution. So in comes mathematicians, and invented a whole new function space called "rapid decay functions", and viola. Just an annecdote, sorry for the snore.

Onward. A modulated Gaussian, is a frequency shifted by $f_0$ Gaussian.
$\large s(t) = cos\({2\pi f_0 t}\)e^{-\frac{1}{2}\({\frac{t - \Delta_t}{\sigma_t/\sqrt{8}}}\)^2}$
It's frequency domain representation is demonstrated below. 
![image](https://github.com/thliebig/openEMS-Project/assets/3908295/fd6c877f-f439-492f-a8d5-36fade2cc763)

Now notice what happens if you choose $f_0$ too low.
![image](https://github.com/thliebig/openEMS-Project/assets/3908295/2b883f6d-9233-4c2f-826d-da84fce56c39)

This basically means that it will be harder for you to filter out the frequency data for the low frequencies, and is best to avoid.

And this is why you want to avoid simulations where your bandwidth covers areas close to $f = 0$. Generally speaking, FDTD is not the tool for these problems, anyhow. 

## **2. High frequency limitations**
In order for the entire frequency response of the Gaussian to be available, the time step needs to be small enough. For example, here is what happens if the time step isn't small enough.
![image](https://github.com/thliebig/openEMS-Project/assets/3908295/9f1a4543-1530-485e-822b-8b7b8c24de2d)

Which is kind of the opposite of trying to cover $f = 0$, but for the high end of the frequency range. You can synthetically increase the sampling frequency by interpolating the time steps, as the data is already lost.

So for this purpose, also choose the time step small enough to suffice for the highest frequency you are trying to cover, and add some, for the Gaussian to decay properly. This, naturally, means longer simualtion time.

## **3. Some stuff about ports**
Here is another issue that I recall, but this is less relevant to **openEMS**, however. At the moment, at least.
When using mode ports, e.g. rectangular waveguides or GCPW mode ports, the fields behave differently for each frequency. Slightly, but still, differently. 

This is an inherent limitation of transient simulations, even in high-end software, so don't sweat it. 


## **Mitigation?**
For the frequency issues, there is none, really. Try to limit the bandwidth of the simualtions to exactly what is necessary. 

I guess that's the answer for the mode port issue, as well. This is very much relevant even for the current implementation of rectangular waveguide ports in openEMS.

**Reply by Sarajevo67** [reply]
Good to see you again,

thanks for your time. I know how it works from physics side. 

My questions was mostly related to the excitation signal bandwidth (f0/2fc), not absolute frequency (fo).
For example, in openEMS, if we choose f0=1 GHz and fc=0.1 GHz, excitation signal length (in timesteps), is twice a length as for f0=1 GHz and fc=0.2 GHz excitation, while timestep is reduced just inverse proportional to fmax increase (1.1/1.2).
As you can see, without any intention to cover frequencies toward f=0, or to dramatically increase fmax, we reduced **simulation time** (almost by half), by increasing BW. That was INITIALLY counterintuitive.

Now, let's try to assess accuracy loss:
**Considering that we _invested_  double numerical mass in half bandwidth simulation, is it reasonable to assume 4x (or sqrt(4)x) higher precision for half excitation BW?**

BR,

DD

**Comment by gadiLhv** [comment]
So I'll focus on the last bit then.

> Now, let's try to assess accuracy loss:
> **Considering that we _invested_ double numerical mass in half bandwidth simulation, is it reasonable to assume 4x (or sqrt(4)x) higher precision for half excitation BW?**

What do you mean by "higher precision for half excitation bandwidth". Precision in what, the S-Parameters deduced? In that sense, only the dispresion effect of the wavguide ports should be taken into account, or at least this is what comes to mind.

**Reply by Sarajevo67** [reply]
I mean that, by longer excitation (in number of TS), for half BW excitation, larger numerical mass is injected into FDTD simulation, which later yields less discretisation noise/uncertainty, for any frequency related port calculation. This is how my general understanding of numerical simulations works. Did I missed something?

**Reply by gadiLhv** [reply]
> which later yields less discretisation noise/uncertainty

Uncertainty and noise are definitely not the same thing. Uncertainty can be reduced by looking for smaller bandwidth, I guess.
The only thing I can think about that can add "noise" is insufficient decay of the signal. Even then, it is more of an interference problem, rather than noise.

>  This is how my general understanding of numerical simulations works

There is definitely more than one kind of numerical simulation. This is specifically about transforming transient results to the frequency domain.

> Did I missed something?

I suggested one mechanism that inserts actual error into the system, which is error in the port **modes** due to wide bandwidth. That is definitely a tradeoff in transient simulation. This is specifically for waveguide ports, however.

**Reply by Sarajevo67** [reply]
`Uncertainty and noise are definitely not the same thing.`

Agree for that , but I said, **discretisation noise**? Gadi, let's not bother on explaining basic things and focus on the issue.

`I suggested one mechanism that inserts actual error into the system`

This is very good idea. For a while, I'm thinking about error propagation modeling.
